File: src/core/files.ts
import { input } from "./gui/dom";

export function open_files(options?: { accept?: string; multiple?: boolean }): Promise<File[]> {
    return new Promise((resolve) => {
        const el = input({ type: "file" });
        el.accept = options?.accept ?? "";
        el.multiple = options?.multiple ?? false;

        el.onchange = () => {
            if (el.files && el.files.length) {
                resolve([...el.files]);
            } else {
                resolve([]);
            }
        };

        el.click();
    });
}

export function read_file(file: File): Promise<ArrayBuffer> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();

        reader.addEventListener("loadend", () => {
            if (reader.result instanceof ArrayBuffer) {
                resolve(reader.result);
            } else {
                reject(new Error("Couldn't read file."));
            }
        });

        reader.readAsArrayBuffer(file);
    });
}

File: webui/src/main/kotlin/world/phantasmal/webui/Files.kt
package world.phantasmal.webui

import kotlinx.browser.document
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.suspendCancellableCoroutine
import org.khronos.webgl.ArrayBuffer
import org.w3c.dom.HTMLInputElement
import org.w3c.dom.asList
import org.w3c.files.File
import org.w3c.files.FileReader

fun openFiles(accept: String = "", multiple: Boolean = false, callback: (List<File>) -> Unit) {
    val el = document.createElement("input") as HTMLInputElement
    el.type = "file"
    el.accept = accept
    el.multiple = multiple

    el.onchange = {
        callback(el.files?.asList() ?: emptyList())
    }

    el.click()
}

@OptIn(ExperimentalCoroutinesApi::class)
suspend fun readFile(file: File): ArrayBuffer = suspendCancellableCoroutine { cont ->
    val reader = FileReader()
    reader.onloadend = {
        if (reader.result is ArrayBuffer) {
            cont.resume(reader.result.unsafeCast<ArrayBuffer>()) {}
        } else {
            cont.cancel(Exception(reader.error.message.unsafeCast<String>()))
        }
    }
    reader.readAsArrayBuffer(file)
}
